### PR TITLE
issue/2824 Switch to rollup from requirejs

### PR DESCRIPTION
**Pull request by oliverfoster**
#2824 
This version does not have the react elements, it is merely to fix the currently broken build process in v5.  

Will provide an almost seamless migration away from build-side [requirejs](https://requirejs.org/docs/optimization.html) and [AMD](https://requirejs.org/docs/whyamd.html) modules to [ES6](https://hacks.mozilla.org/2015/08/es6-in-depth-modules/) modules.  

#### Facilitates
* future proof
  * `import` and `export` directives
  * beyond ES8 code
* development improvements
  * much faster compilation and recompilation times on all grunt tasks, this is as a cache is now generated for the build process
  * supports larger code bases
  * move towards supporting intellisense in code editors
* backward-compatibility
  * ES6 modules are reregistered in the browser as AMD modules, so `var Adapt = require('core/js/adapt');` etc will still work in the console and for other plugins
  * we can now rework Adapt to ES6 modules whilst retaining support for the existing AMD modules
  * still ie11 compatible

#### Downsides
* External modules such as [libraries/mediaelement-fullscreen-hook.js](https://github.com/adaptlearning/adapt-contrib-media/blob/master/libraries/mediaelement-fullscreen-hook.js), which are required by internal modules and require internal modules are now considered circular dependencies and will break the loading process. A temporary fix is included until that module can be moved back into the `js` folder
* Jumping to class function line numbers from the code inspector in the debugger is out of order as the define/require directive conversions moves the line numbers
* Currently this will only work with transpilation back to ES5 as calling an ES6 `constructor` function without the `new` keyword is not allowed - which is an essential part of the way Backbone constructs classes.

#### Added
* [rollup](https://rollupjs.org/guide/en/) as the bundler
* [babel](https://babeljs.io/) is used as the compiler/transpiler through the [rollup plugin](https://www.npmjs.com/package/@rollup/plugin-babel)
* [babel-plugin-transform-amd-to-es6](https://github.com/cgkineo/babel-plugin-transform-amd-to-es6) (custom made) to convert [AMD](https://requirejs.org/docs/whyamd.html) modules to [ES6](https://hacks.mozilla.org/2015/08/es6-in-depth-modules/) modules
* rollup compilation cache lives at `<%= outputdir %>/.cache` to speed up compilation times
* [jsconfig.json](https://code.visualstudio.com/docs/languages/jsconfig) to support intellisense
* grunt flag `--disable-cache` to disable the cache and recompile from scratch

#### Removed
* [requirejs](https://requirejs.org/docs/optimization.html) as bundler in [grunt](https://gruntjs.com/)
* [babel grunt task](https://github.com/babel/grunt-babel) is now integrated into the rollup process
* requirejs `config.js` file has been moved to the `javascript` grunt task config

#### Fixed
* `QuestionView` has two different `return` statements, so cannot resolve to one `export default` directive

#### Testing
* Use master of all plugins by running:
```js
git clone https://github.com/adaptlearning/adapt_framework 2824
cd 2824
git checkout issue/2824
adapt devinstall && npm install
grunt dev
```

#### Example
Shows fast compilation times, `import`/`export`, debugging and `require` functioning with an ES6 module in the client-side:

rollup-es-modules2

### CODE REVIEW

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

File: src/core/js/adapt.js [hunk 1]
```diff
@@ -1,431 +1,429 @@
...
+   * @return {string} One of the following (or `undefined` if not found):
+   * - "course"
+   * - "contentObjects"
+   * - "blocks"
+   * - "articles"
+   * - "components"
+   */
+  mapById(id) {
+    return this.data.mapById(id);
+  }
```
Comment (moloko) [inline src/core/js/adapt.js:324]: suggest either removing this or at least adding a deprecated warning

https://github.com/adaptlearning/adapt_framework/issues/2908

Comment (oliverfoster) [reply src/core/js/adapt.js:324]: Can I do that in a different pr?

Comment (oliverfoster) [reply src/core/js/adapt.js]: Have removed.

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

**Comment by chris-steele** [comment]
Looks fantastic, only issue I saw was a `childViews` reference runtime error with master branch of [adapt-contrib-boxMenu](https://github.com/adaptlearning/adapt-contrib-boxMenu), but works fine using latest release of said menu. :+1:

**Review by chris-steele — APPROVED** [review]

**Comment by oliverfoster** [comment]
@moloko Could you +1 when you get a mo?

**Pull request merged**